lib/grandpa/engine.go

```diff
@@ -0,0 +1,475 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+type ephemeralService interface {
+	Start() error
+	Stop() error
+}
+
+type finalizationHandler struct {
+	servicesLock       sync.Mutex
+	finalizationEngine ephemeralService
+	votingRound        ephemeralService
+
+	servicesBuilder func() (engine, voting ephemeralService)
+	timeoutStop     time.Duration
+	initiateRound   func() error
+	observableErrs  chan error
+	stopCh          chan struct{}
+	handlerDone     chan struct{}
+}
+
+func newFinalizationHandler(service *Service) *finalizationHandler {
+	// builder is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	builder := func() (engine, voting ephemeralService) {
+		finalizationEngine := newFinalizationEngine(service)
+		votingRound := newHandleVotingRound(service, finalizationEngine.actionCh)
+		return finalizationEngine, votingRound
+	}
+
+	return &finalizationHandler{
+		servicesBuilder: builder,
+
+		timeoutStop:    5 * time.Second,
+		initiateRound:  service.initiateRound,
+		observableErrs: make(chan error),
+		stopCh:         make(chan struct{}),
+		handlerDone:    make(chan struct{}),
+	}
+}
+
+func (fh *finalizationHandler) Start() (errsCh <-chan error, err error) {
+	go fh.runFinalization()
+	return fh.observableErrs, nil
+}
+
+func (fh *finalizationHandler) stop() (err error) {
+	stopWg := new(sync.WaitGroup)
+	stopWg.Add(2)
+
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	errorsCh := make(chan error, 2)
+	defer close(errorsCh)
+
+	go func() {
+		defer stopWg.Done()
+		errorsCh <- fh.finalizationEngine.Stop()
+	}()
+	go func() {
+		defer stopWg.Done()
+		errorsCh <- fh.votingRound.Stop()
+	}()
```

Comment:
How do you know which service returned which error?

Reply:
Changing to an `errs := [2]error{}` and now we can check index 0 for first service, index 1 for the other service, similar to here -> https://github.com/timwu20/service-example/blob/510df34a7051daf9f88023b81ec99c1e827d0955/service/sub_service_round.go#L102